RealmSwift/Realm.swift

```diff
@@ -366,6 +370,115 @@ import Realm.Private
         return rlmRealm.inWriteTransaction
     }
 
+#if REALM_ASYNC_WRITES
+// MARK: Asynchronous Transactions
+
+    /**
+     Asynchronously performs actions contained within the given block inside a write transaction.
+     The write transaction is begun asynchronously as if calling `beginAsyncWrite`,
+     and by default the transaction is commited asynchronously after the block completes.
+     You can also explicitly call `commitWrite` or `cancelWrite` from
+     within the block to synchronously commit or cancel the write transaction.
+
+     @param block The block containing actions to perform.
+
+     @param completionBlock A block which will be called on the source thread or queue
+                        once the commit has either completed or failed with an error.
+
+     @return An id identifying the asynchronous transaction which can be passed to
+             `cancelAsyncWrite` prior to the block being called to cancel
+             the pending invocation of the block.
+    */
+    @discardableResult
+    public func writeAsync(_ block: @escaping (AsyncTransactionId) -> Void, _ onComplete: ((Swift.Error?) -> Void)? = nil) throws -> AsyncTransactionId {
+        return try beginAsyncWrite { asyncTransactionId in
+            block(asyncTransactionId)
+            commitAsyncWrite(onComplete)
+        }
+    }
+
+    /**
+     Begins an asynchronous write transaction.
+     This function asynchronously begins a write transaction on a background
+     thread, and then invokes the block on the original thread or queue once the
+     transaction has begun. Unlike `beginWrite`, this does not block the
+     calling thread if another thread is current inside a write transaction, and
+     will always return immediately.
+     Multiple calls to this function (or the other functions which perform
+     asynchronous write transactions) will queue the blocks to be called in the
+     same order as they were queued. This includes calls from inside a write
+     transaction block, which unlike with synchronous transactions are allowed.
+
+     @param asyncWriteBlock The block containing actions to perform inside the write transaction.
+            `asyncWriteBlock` should end by calling `commitAsyncWrite` or `commitWrite`.
+            Returning without one of these calls is equivalent to calling `cancelAsyncWrite`.
+
+     @return An id identifying the asynchronous transaction which can be passed to
+             `cancelAsyncWrite` prior to the block being called to cancel
+             the pending invocation of the block.
+     */
+    @discardableResult
+    public func beginAsyncWrite(_ asyncWriteBlock: @escaping (AsyncTransactionId) -> Void) throws -> AsyncTransactionId {
```

Comment:
This doesn't appear to ever actually throw an error and so shouldn't be `throws`.